vsphere/internal/virtualdevice/virtual_machine_cdrom_subresource.go

```diff
@@ -0,0 +1,507 @@
+	// as new devices.
+	for n, device := range devices {
+		m := make(map[string]interface{})
+		vd := device.GetVirtualDevice()
+		ctlr := l.FindByKey(vd.ControllerKey)
+		if ctlr == nil {
+			return fmt.Errorf("could not find controller with key %d", vd.Key)
+		}
+		m["key"] = int(vd.Key)
+		m["device_address"] = computeDevAddr(vd, ctlr.(types.BaseVirtualController))
+		r := NewCdromSubresource(c, d, m, nil, n)
+		if err := r.Read(l); err != nil {
+			return fmt.Errorf("%s: %s", r.Addr(), err)
+		}
+		newSet = append(newSet, r.Data())
+	}
+
+	log.Printf("[DEBUG] CdromRefreshOperation: Resource set to write after adding orphaned devices: %s", subresourceListString(newSet))
+	log.Printf("[DEBUG] CdromRefreshOperation: Refresh operation complete, sending new resource set")
+	return d.Set(subresourceTypeCdrom, newSet)
+}
+
+// CdromPostCloneOperation normalizes CDROM devices on a freshly-cloned virtual
+// machine and outputs any necessary device change operations. It also sets the
+// state in advance of the post-create read.
+//
+// This differs from a regular apply operation in that a configuration is
+// already present, but we don't have any existing state, which the standard
+// virtual device operations rely pretty heavily on.
+func CdromPostCloneOperation(d *schema.ResourceData, c *govmomi.Client, l object.VirtualDeviceList) (object.VirtualDeviceList, []types.BaseVirtualDeviceConfigSpec, error) {
+	log.Printf("[DEBUG] CdromPostCloneOperation: Looking for post-clone device changes")
+	// While we are currently only restricting CD devices to one device, we have
+	// to actually account for the fact that someone could add multiple CD drives
+	// out of band. So this workflow is similar to the multi-device workflow that
+	// exists for network devices.
+	devices := l.Select(func(device types.BaseVirtualDevice) bool {
+		if _, ok := device.(*types.VirtualCdrom); ok {
+			return true
+		}
+		return false
+	})
+	log.Printf("[DEBUG] CdromPostCloneOperation: CDROM devices located: %s", DeviceListString(devices))
+	curSet := d.Get(subresourceTypeCdrom).([]interface{})
+	log.Printf("[DEBUG] CdromPostCloneOperation: Current resource set from configuration: %s", subresourceListString(curSet))
+	var srcSet []interface{}
+
+	// Populate the source set as if the devices were orphaned. This give us a
+	// base to diff off of.
+	log.Printf("[DEBUG] CdromPostCloneOperation: Reading existing devices")
+	for n, device := range devices {
+		m := make(map[string]interface{})
+		vd := device.GetVirtualDevice()
+		ctlr := l.FindByKey(vd.ControllerKey)
+		if ctlr == nil {
+			return nil, nil, fmt.Errorf("could not find controller with key %d", vd.Key)
+		}
+		m["key"] = int(vd.Key)
+		m["device_address"] = computeDevAddr(vd, ctlr.(types.BaseVirtualController))
+		r := NewCdromSubresource(c, d, m, nil, n)
+		if err := r.Read(l); err != nil {
+			return nil, nil, fmt.Errorf("%s: %s", r.Addr(), err)
```

Comment:
This will actually report a much more verbose error from the `Read` function. It will look something like:

```
vsphere_virtual_machine.vm: cdrom.0: Something bad happened in the read function
```